### DISCUSSION TITLE
I am not able to compile 'BLEMIDI-Adapter.ino' on Raspberry PICO W

### DISCUSSION THREAD

**Discussion by maarthome** [tttapa/Control-Surface#1014] [Q&A]
I imported the raspberry pico w library (https://github.com/earlephilhower/arduino-pico/releases/download/global/package_rp2040_index.json)
But unfortunately still do not get it working..... :-( 
Any help is welcome :-)
Thanks!

Error message:

C:\Users\maart\Documents\Arduino\libraries\Control_Surface\examples\3. MIDI Interfaces\BLEMIDI-Adapter\BLEMIDI-Adapter.ino:51:1: error: 'BluetoothMIDI_Interface' does not name a type
   51 | BluetoothMIDI_Interface midi_ble;
      | ^~~~~~~~~~~~~~~~~~~~~~~
C:\Users\maart\Documents\Arduino\libraries\Control_Surface\examples\3. MIDI Interfaces\BLEMIDI-Adapter\BLEMIDI-Adapter.ino: In function 'void setup()':
C:\Users\maart\Documents\Arduino\libraries\Control_Surface\examples\3. MIDI Interfaces\BLEMIDI-Adapter\BLEMIDI-Adapter.ino:59:3: error: 'midi_ble' was not declared in this scope
   59 |   midi_ble.setName("MIDI Adapter");
      |   ^~~~~~~~
C:\Users\maart\Documents\Arduino\libraries\Control_Surface\examples\3. MIDI Interfaces\BLEMIDI-Adapter\BLEMIDI-Adapter.ino: In function 'void loop()':
C:\Users\maart\Documents\Arduino\libraries\Control_Surface\examples\3. MIDI Interfaces\BLEMIDI-Adapter\BLEMIDI-Adapter.ino:73:29: error: 'midi_ble' was not declared in this scope
   73 |   digitalWrite(LED_BUILTIN, midi_ble.isConnected() ? HIGH : LOW);
      |                             ^~~~~~~~

Using library Control Surface at version 2.0.0 in folder: C:\Users\maart\Documents\Arduino\libraries\Control_Surface 
Using library SPI at version 1.0 in folder: C:\Users\maart\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.7.2\libraries\SPI 
exit status 1

Compilation error: 'BluetoothMIDI_Interface' does not name a type

**Comment by tttapa** [comment]
Did you enable the Bluetooth stack as described in https://tttapa.github.io/Control-Surface-doc/Doxygen/db/d99/md_pages_MIDI-over-BLE.html?

**Reply by maarthome** [reply]
Hi Pieter,

Thanks a lot for the fast reply. I went to bed after so many hours of staring and compiling and trying.... 
I read the BLE related documentation and I selected indeed 'IPv4 + Bluetooth' in the 'Tools' menu....

This morning I uninstalled the Arduino IDE and all files that are left are deinstallation and installed everything from scratch.
And guess what.... BLE compiles now :-) I am happy :-)

The new adventure is making a simple USB to BLE pass-true sketch. Using 'USBMIDI-Adapter.ino' and replacing all related to the serial interface with the BLE stuff from other sketches....
The BLE works but the USB is not shown in my DAW. ( I use 'Adafruit TinyUSB'. which normally compiles OK)

Do you have a suggestion?
I feel somewhat ashamed to bother you.... I am used to programming MATLAB code is a 'rechttoe rechtaan' manner... not object oriented.
Also, I am VERY impressed with all the code you have made in all these last years. Very well documented!! It covers an enourmous parameter/possibility space. Chapeau!! 
Thanks for making this available for others with less experience.

Enjoy your weekend!

**Reply by tttapa** [reply]
Thanks for letting me know! You're right, there seems to be some unwanted interference between the BLE MIDI code and the Adafruit TinyUSB MIDI code: if the BLE MIDI interface is initialized first, then the Adafruit TinyUSB library does not seem to get initialized.

The following works for me:

```cpp
#include <Control_Surface.h>
#include <MIDI_Interfaces/BluetoothMIDI_Interface.hpp>

// Instantiate a MIDI over USB interface
USBMIDI_Interface midi_usb;
// Instantiate a MIDI over BLE interface
BluetoothMIDI_Interface midi_ble;
// Instantiate the pipe to connect the two interfaces
BidirectionalMIDI_Pipe pipes;

void setup() {
  // Change the name of the BLE device (must be done before initializing it)
  midi_ble.setName("MIDI Adapter");
  // Manually route MIDI input from the USB interface to the BLE interface,
  // and the MIDI input from the BLE interface to the USB interface
  midi_usb | pipes | midi_ble;
  // Initialize the MIDI interfaces
  MIDI_Interface::beginAll();
  // Initialize the built-in LED
  pinMode(LED_BUILTIN, OUTPUT);
}

void loop() {
  // Continuously poll all interfaces and route the traffic between them
  MIDI_Interface::updateAll();
  // Display the connection status using the built-in LED
  digitalWrite(LED_BUILTIN, midi_ble.isConnected() ? HIGH : LOW);
}
```
But swapping the order of the MIDI and BLE interfaces, only BLE works. I'll investigate further.

---

**Edit**: I've opened an issue here: https://github.com/earlephilhower/arduino-pico/issues/2022

**Reply by maarthome** [reply]
Hi Pieter,

It works now!! :-) Makes me happy!
This is the start of exploring a new universe for me. Controlling/manipulating MIDI devices... synths and samplers.
Input will be all possible human interaction.... sound, gestures, movements , etc, etc... perhaps even brain activity expressed in voltage....

Thanks so VERY VERY MUCH!
Maarten

**Reply by tttapa** [reply]
Glad to hear, enjoy!

One other thing to keep in mind when using MIDI over USB on the Pico is https://github.com/adafruit/Adafruit_TinyUSB_Arduino/issues/238
